Client/Frontend/Browser/LoginsHelper.swift

```diff
@@ -203,36 +218,37 @@ class PasswordHelper: BrowserHelper {
 
         if let credential = credential {
             // If we have some credentials, we'll show a prompt with them.
-            let password = Password(credential: credential, protectionSpace: challenge.protectionSpace)
-            promptForUsernamePassword(viewController, password: password, completion: completion)
-        } else {
-            // Otherwise, try to look one up
-            let options = QueryOptions(filter: challenge.protectionSpace.host, filterType: .None, sort: .None)
-            profile.passwords.get(options, complete: { (cursor) -> Void in
-                var password = cursor[0]
-                if password == nil {
-                    password = Password(credential: nil, protectionSpace: challenge.protectionSpace)
-                }
-                self.promptForUsernamePassword(viewController, password: password!, completion: completion)
-            })
+            return promptForUsernamePassword(viewController, credentials: credential, protectionSpace: challenge.protectionSpace)
+        }
+
+        // Otherwise, try to look one up
+        let options = QueryOptions(filter: challenge.protectionSpace.host, filterType: .None, sort: .None)
+
+        return profile.logins.getLoginsForProtectionSpace(challenge.protectionSpace).bindQueue(dispatch_get_main_queue()) { res in
+            var login: LoginData? = nil
+            if let cursor = res.successValue {
+                login = cursor[0]
+            }
```

Comment:
If we're fine muffling failure and instead prompting (and I agree that that's the sanest thing to do), we can simplify this:

```
        { res in
    let credentials = res.successValue?[0]?.credentials
    return self.promptForUsernamePassword(…, credentials: credentials, …)
}
```